### PR TITLE
DataSource: show the uid in edit url, not the local id

### PR DESCRIPTION
**Pull request by ryantxu**
Rather than showing local datasource id in the URL, we should use the uid.  See also:
https://github.com/grafana/grafana/pull/33817

image

### CODE REVIEW

File: public/app/features/datasources/state/actions.ts [hunk 1]
```diff
@@ -111,9 +111,18 @@ export function loadDataSources(): ThunkResult<void> {
...
-export function loadDataSource(id: number): ThunkResult<void> {
+export function loadDataSource(uid: string): ThunkResult<void> {
   return async (dispatch) => {
-    const dataSource = (await getBackendSrv().get(`/api/datasources/${id}`)) as DataSourceSettings;
+    // Check if this is an old id format!
+    const id = parseInt(uid, 10);
+    if (id) {
+      const ds = (await getBackendSrv().get(`/api/datasources/${id}`)) as DataSourceSettings;
+      if (ds?.uid) {
+        console.log('TODO... redirect!', ds.uid);
```
Comment (ryantxu) [inline public/app/features/datasources/state/actions.ts:121]: must be something cleaner than this!  any ideas?

Comment (torkelo) [reply public/app/features/datasources/state/actions.ts:121]: I  can try figure this out, pushing something soon

Comment (ryantxu) [reply public/app/features/datasources/state/actions.ts]: just saw this -- pushed somehting but it breaks with:
image

**Review by torkelo — COMMENTED** [review]
Pushed an update to the load data source & redirect handling on data source settings page

File: public/app/features/datasources/state/actions.ts (hunk 1 above)
Comment (torkelo) [reply public/app/features/datasources/state/actions.ts]: it's working now, was a bit tricky to fix

**Comment by ryantxu** [comment]
e2e tests are now passing -- I think it is ready to go

**Review by torkelo — APPROVED** [review]

**Comment by Clarity-89** [comment]
This breaks quite a lot of things in the Enterprise, particularly caching, which is set up to retrieve cache data by datasource id: https://github.com/grafana/grafana-enterprise/blob/main/src/public/caching/state/actions.ts. Also for some reason, I get the uid for all datasources as `x`.

File: public/app/features/datasources/state/actions.ts [hunk 2]
```diff
@@ -123,6 +125,50 @@ export function loadDataSource(id: number): ThunkResult<void> {
   };
 }
 
+/**
+ * Get data source by uid or id, if old id detected handles redirect
+ */
+async function getDataSourceUsingUidOrId(uid: string): Promise<DataSourceSettings> {
```
Comment (Clarity-89) [inline public/app/features/datasources/state/actions.ts:131]: Wouldn't it be easier to allow both string and number here and then make API call according to the type? Or is it not always given that `id` will be a number and `uid` a string?

**Pull request merged**